pkg/client/block/client.go

```diff
@@ -0,0 +1,169 @@
+package block
+
+import (
+	"context"
+	"fmt"
+	"log"
+
+	"cosmossdk.io/depinject"
+
+	"pocket/pkg/client"
+	"pocket/pkg/either"
+	"pocket/pkg/observable"
+	"pocket/pkg/observable/channel"
+)
+
+var (
+	_ client.BlockClient = (*blockClient)(nil)
+	_ client.Block       = (*cometBlockEvent)(nil)
+)
+
+// blockClient implements the BlockClient interface.
+type blockClient struct {
+	// endpointURL is the URL of RPC endpoint which eventsClient subscription
+	// requests will be sent.
+	endpointURL string
+	// eventsClient is the events query client which is used to subscribe to
+	// newly committed block events. It emits an either value which may contain
+	// an error, at most, once and closes immediately after if it does.
+	eventsClient client.EventsQueryClient
+	// latestBlockObsvblsReplay is a replay observable with replay buffer size 1,
+	// which holds the "active latest block observable" which is notified when
+	// block commit events are received by the events query client subscription
+	// created in goPublishBlocks. This observable (and the one it emits) closes
+	// when the events bytes observable returns an error and is updated with a
+	// new "active" observable after a new events query subscription is created.
+	latestBlockObsvblsReplay observable.ReplayObservable[client.BlocksObservable]
+	// latestBlockObsvblsReplayPublishCh is the publish channel for latestBlockObsvblsReplay.
+	// It's used to set blockObsvbl initially and subsequently update it, for
+	// example, when the connection is re-established after erroring.
+	latestBlockObsvblsReplayPublishCh chan<- client.BlocksObservable
+}
+
+func NewBlockClient(
+	ctx context.Context,
+	deps depinject.Config,
+	cometWebsocketURL string,
+) (client.BlockClient, error) {
+	// Initialize block client
+	bClient := &blockClient{endpointURL: cometWebsocketURL}
+	bClient.latestBlockObsvblsReplay, bClient.latestBlockObsvblsReplayPublishCh =
+		channel.NewReplayObservable[client.BlocksObservable](ctx, 1)
+
+	// Inject dependencies
+	if err := depinject.Inject(deps, &bClient.eventsClient); err != nil {
+		return nil, err
+	}
+
+	// Concurrently publish blocks to the observable emitted by latestBlockObsvblsReplay.
+	go bClient.goPublishBlocks(ctx)
+
+	return bClient, nil
+}
+
+// CommittedBlocksSequence returns a ReplayObservable, with a replay buffer size
+// of 1, which is notified when block commit events are received by the events
+// query subscription.
+func (bClient *blockClient) CommittedBlocksSequence(ctx context.Context) client.BlocksObservable {
+	replayedBlocksObservable := bClient.latestBlockObsvblsReplay.Last(ctx, 1)[0]
```

Comment:
use local constant here

Reply:
**TL;DR: I'm less convinced on this one. I #PUC'd it up. :stuck_out_tongue_closed_eyes:**

I agree with moving the `1` to a constant where it's used as the `replayBufferSize` argument but I'm gonna push back on this one. In the former, we're considering "how many observable references should we keep". Here, we're considering "how many observable references should we pull out" in a context where we're expected to return a single observable. I updated the comment:

```go
	// Get the latest block observable from the replay observable. We only ever
	// want the last 1 as any prior latest block observable values are closed.
	// Directly accessing the zeroth index here is safe because the call to Last
	// is guaranteed to return a slice with at least 1 element.
	return bClient.latestBlockObsvblsReplay.Last(ctx, 1)[0]
```

Reply:
Ty for the explanation!